README.md
# Interactive Data Acoustic Modeling

Welcome to the Interactive Data Analysis Acoustic Modeling Platform, a comprehensive program designed to facilitate the analysis, cleaning, and modeling of audio data. This project aims to provide a user-friendly interface for importing data from various sources, cleaning it by handling missing values and inconsistencies, and performing in-depth data analysis to uncover valuable insights. The toolkit excels in handling data formatting issues, such as converting audio files from multiple different formats to wav format, ensuring seamless integration into your analysis pipeline.

This platform is designed to enable users to seamlessly import, clean, visualize, analyze, and model data
using scientific calculations and data visualization tools.

To start, simply press the "Open a File" button on our platform's GUI and select a sound file (.wav,.mp3,.m4a,.aac) from your files. Then press the "Analyze File" button to have the file analyzed and processed and spit out the data onto your screen. This data includes the files length, the resonant frequency and the difference in seconds. Now you can select the different graphs to view from the buttons under the graph. The "Intensity Graph" button displays the intensity graph, the "Waveform Graph" button displays the waveform graph, the "Combine RT60 Graphs" shows the 3 RT60 graphs in one graph, finally the "Cycle RT60 Graphs" button goes through the low, mid and high RT60 graphs.

CHANGELOG.md
# Changelog:
# 12/12/23
- Updated README.md
- Updated CHANGELOG.md
- Updated pyproject,toml
- Create requirements.txt
- Updated requirements.txt
- Fixed bad Comments on code
# 12/11/23
- Added CHANGELOG.md, soundfiles of the recording, and updated README.md a little
- Update CHANGELOG.md
- Added a pyproject.toml
- Found resonance and RT60 time difference. Also removed unnecessary value displays.
- UI overhaul. Specific buttons for specific actions. General cleanup of code.
# 12/10/23
- Got the 3 RT60 graphs working
- Converted main file to mvc format
- GUI update, Swap Graphs Button, RT60 Info
- Fixed issue with frequency always showing as 0
- Made changes to how files are processed and cleaned. Cleaned.wav created.
- Fix 2chan files not being properly converted
# 12/9/23
- Added first frequency graph.
- Added conversion to wav and ability to take in multiple files.
- Added another graph for RT60 at high frequency with fixed comments.
- Fixed an issue with previous graph data lingering when selecting another file to analyze. Also organized the code a little bit.
# 12/08/23
- Fixed an issue where labels were being created on top of each other instead of updating one label.
- Fixed merging issues
# 12/07/23
- Added working "Clean File" button, modified "analyze_file" to properly output number of channels.
# 12/06/23
- Added recorded wav file and merged draft branch
# 12/04/23
- Began coding mp4 to wav file conversion
# 12/03/23
- Added README.md
- Changed ui to show graph and info
- Added a program that takes in a sound file to clean and convert it to .wav as needed. Uses ffmpeg, Currently not fully functional
# 12/02/23
- Repository created


model.py
#modle.py
from pydub import AudioSegment
import os
import tempfile
import tkinter as tk
from tkinter import filedialog as fd
import ffmpeg
import wave
from scipy.io import wavfile
from scipy.signal import welch
import soundfile as sf
import numpy as np
import matplotlib.pyplot as plt


class Model:
    # Initialize variable for calculations
    def __init__(self):       
        self.samplerate = 0
        self.data = None
        self.channels = 0
        self.freqs = None
        self.spectrum = None
        self.t = None
        self.im = None

    # Read the audio file
    def read_audio(self, file_path):
        try:
            audio = AudioSegment.from_file(file_path)
            return audio
        except Exception as e:
            print(f"Error loading audio file: {e}")
            return None

    # Convert the audio file to .wav
    def convert_to_wav(self, audio_data, output_path):
        # Convert to wav format
        if format != "wav":
            audio_data.export(output_path, format="wav")

    # Read the newly created w.av file
    def read_wav(self, wav_path):
        try:
            wav = AudioSegment.from_file(wav_path)
            return wav
        except Exception as e:
            print(f"Error loading wav file: {e}")
            return None

    # Convert the file to one channel
    def mono_channel(self, wav_file, output_path):
        wav_file = wav_file.set_channels(1)
        wav_file.export(output_path, format="wav")       
            
    # Remove metadat from the file using os
    def remove_metadata(self, wav_file, output_path):
        # Use ffmpeg to remove metadata
        os.system(f'ffmpeg -i "{wav_file}" -map_metadata -1 -acodec copy "{output_path}"')

    # Analyze the file and grab data
    def analyze_file(self, file_path):
        wav_fname = file_path
        with wave.open(wav_fname, 'rb') as wav_read:
            self.channels = wav_read.getnchannels()
        self.samplerate, self.data = wavfile.read(wav_fname)
        self.spectrum, self.freqs, self.t, self.im = plt.specgram(self.data, Fs=self.samplerate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))

    # Calculate the resonance
    def calculate_res(self, file_path):
        frequencies, power = welch(self.data, self.samplerate, nperseg=4096)
        dominant = frequencies[np.argmax(power)]
        return dominant
    
    # Calulate RT60 values
    def RT60(self, low_freq, high_freq):
        # Find indices of frequencies within the specified range
        indices = np.where((self.freqs >= low_freq) & (self.freqs <= high_freq))[0]
        
        # Find the frequency with maximum intensity within the specified range
        self.target_frequency_index = indices[np.argmax(np.max(self.spectrum[indices, :], axis=1))]

        
        data_in_db = self.frequency_check(self.freqs[self.target_frequency_index])

        value_of_max_less_5 = data_in_db[self.target_frequency_index] - 5
        value_of_max_less_5 = self.find_nearest_value(data_in_db, value_of_max_less_5)
        index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

        value_of_max_less_25 = data_in_db[self.target_frequency_index] - 25
        value_of_max_less_25 = self.find_nearest_value(data_in_db, value_of_max_less_25)
        index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

        # RT20 value
        rt20 = (self.t[index_of_max_less_5] - self.t[index_of_max_less_25])[0]

        # Calculating the RT60 value
        self.rt60 = 3 * rt20

        return abs(self.rt60)

    # Calculate the RT60 dif
    def RT60_dif(self, low, mid, high):
        RT60avg = (low + mid + high) / 3
        dif = RT60avg - .5
        return abs(round(dif,2))


    # Checking for frequency
    def frequency_check(self, target_frequency):
        
        # Identify a frequency to check
        index_of_frequency = np.where(self.freqs == target_frequency)[0][0]
        # Find sound data for a particular frequency
        data_for_frequency = self.spectrum[index_of_frequency]
        # Change a digital signal for a value in decibels
        data_in_db_fun = 10 * np.log10(data_for_frequency)
        return data_in_db_fun

    # Find the nearest values for plotting
    def find_nearest_value(self, array, value):
        array = np.asarray(array)
        idx = (np.abs(array - value)).argmin()
        return array[idx]
    



        


controller.py
#controller.py
import tkinter as tk
from tkinter import ttk
from tkinter import filedialog as fd
import matplotlib.pyplot as plt
import os
from view import View
from model import Model
import numpy as np

class Controller:
    # Initialize tools
    def __init__(self, model, view):
        self.model = model
        self.view = view
        
        self.root = self.view.root
        
        self.output_path = "cleaned.wav"
        self.setup_callbacks()

    #  Process the audio using model functions
    def process_audio(self, output_path):

        
        audio_data = self.model.read_audio(self.view.gfile)
        self.model.convert_to_wav(audio_data, output_path)

        
        wav_data = self.model.read_wav(output_path)
        self.model.mono_channel(wav_data, output_path)
        self.model.remove_metadata(wav_data, output_path)

    # Setup button calls
    def setup_callbacks(self):
        self.view.open_button.config(command=self.select_file)
        self.view.analyze_button.config(command=self.analyze_file)

    # Select file function
    def select_file(self):
        filetypes = (
            ('Sound files', '*.wav *.m4a *.aac *.mp3'),
            ('All files', '*.*')
        )

        filename = fd.askopenfilename(
            title='Open a file',
            initialdir='/',
            filetypes=filetypes)

        # Set view gfile
        self.view.gfile = filename

        # Set justname to only the name of the file and not the whole path
        justname = os.path.basename(filename)

        # Update the view
        self.view.gfile_label.config(text=f"File Name: {justname}")


        
        # Tkinter Analyze button set back up
        self.view.analyze_button = ttk.Button(
            self.root,
            text='Analyze File',
            command=self.analyze_file
        )
        self.view.analyze_button.grid(column=2, row=1)


    def analyze_file(self):
        # Call the analyze_file method from the model
        self.process_audio(self.output_path)
        
        self.model.analyze_file(self.output_path)

        # Call the plot_waveform method from the view
        self.view.plot_waveform()

        # Display resonance
        res = round(self.model.calculate_res(self.output_path),2)
        self.view.res_label.config(text=f"Resonant Frequency: {res} Hz")

        # Low
        low = self.model.RT60(20,250)
        # Mid
        mid = self.model.RT60(250,1000)
        # High
        high = self.model.RT60(1000,5000)

        # Display RT60 difference
        RT60DIF = self.model.RT60_dif(low,mid,high)
        self.view.RT60dif_label.config(text=f"Difference: {RT60DIF}s")

        # Set count
        self.view.count = 0

        # Make analyze button disappear
        self.view.analyze_button.grid_forget()

        # Make graph controls appear
        self.view.swap_graphs_button.grid(column=2, row=5, columnspan=2)
        self.view.combine_button.grid(column=2, row=7, columnspan=2)
        self.view.waveform_graph_button.grid(column=1, row=5, columnspan=2)
        self.view.intensity_graph_button.grid(column=0, row=5, columnspan=2)

        


pydub_possible_importer.py
from pydub import AudioSegment
import os
import tempfile
import ffmpeg


def read_audio(file_path):
    try:
        audio = AudioSegment.from_file(file_path)
        return audio
    except Exception as e:
        print(f"Error loading audio file: {e}")
        return None

def clean_audio(audio):
    # Convert to one channel if multichannel
    if audio.channels > 1:
        audio = audio.set_channels(1)

    return audio

def convert_to_wav(audio, output_path):
    # Convert to wav format
    if format != "wav":
        audio.export(output_path, format="wav")

def remove_metadata(input_path, output_path):
    # Use ffmpeg to remove metadata
    os.system(f'ffmpeg -i "{input_path}" -map_metadata -1 -acodec copy "{output_path}"')

if __name__ == "__main__":
    input_file_path = "file.m4a"  # Replace with the path to your audio file
    output_file_path = "cleaned.wav"  # Replace with the desired output path

    # Step 1: Load the audio
    audio_data = read_audio(input_file_path)

    if audio_data:
        # Step 2: Clean the audio data
        cleaned_audio = clean_audio(audio_data)

        # Step 3: Save cleaned audio to temporary wav file
        temp_wav_file = tempfile.NamedTemporaryFile(suffix=".wav", delete=False)
        convert_to_wav(cleaned_audio, temp_wav_file.name)

        # Step 4: Remove metadata from the temporary wav file
        remove_metadata(temp_wav_file.name, output_file_path)

        # Step 5: Print the cleaned audio information
        print("Cleaned Audio Information:")
        print(cleaned_audio)
    else:
        print("Audio file not loaded.")


view.py
# view.py
import tkinter as tk
from tkinter import ttk
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import matplotlib.pyplot as plt
import numpy as np
import os


class View:
    # Sets up tkinter, widgets, and variables
    def __init__(self, model):
        self.root = tk.Tk()
        self.root.title('Interactive Data Acoustic Modeling')
        self.root.resizable(False, False)
        self.root.geometry('700x800')
        
        self.model = model
        
        self.create_widgets()
       
        self.target_frequency_index = 0
        self.rt60 = 0

    def create_widgets(self):
        
        # Labels
        self.gfile_label = ttk.Label(self.root, text="File Name:")
        self.gfile_label.grid(column=1, row=3, columnspan=2)

        

        self.length_label = ttk.Label(self.root, text=f"File Length = 0s")
        self.length_label.grid(column=1, row=6, columnspan=2)

        

        self.res_label = ttk.Label(self.root, text="Resonant Frequency: ____ Hz")
        self.res_label.grid(column=1, row=7, columnspan=2)

        self.RT60dif_label = ttk.Label(self.root, text="Difference: __.__s")
        self.RT60dif_label.grid(column=1, row=8, columnspan=2)

        self.message_label = ttk.Label(self.root, text="")
        self.message_label.grid(column=1, row=12, columnspan=2)

        # Tkinter Open button
        self.open_button = ttk.Button(
            self.root,
            text='Open a File',
            command=self.select_file
        )
        self.open_button.grid(column=1, row=1)

        # Tkinter Analyze button
        self.analyze_button = ttk.Button(
            self.root,
            text='Analyze File',
            command=self.analyze_file
        )
        self.analyze_button.grid(column=2, row=1)

        self.analyze_button.grid_forget()

        
        
        # Tkinter Cycle Freq Graphs button
        self.swap_graphs_button = ttk.Button(
            self.root,
            text='Cycle RT60 Graphs',
            command=self.swap_RT60_graphs
        )
        self.swap_graphs_button.grid(column=2, row=5, columnspan=2)

        self.swap_graphs_button.grid_forget()

        # Tkinter Combine Freq Graphs button
        self.combine_button = ttk.Button(
            self.root,
            text='Combine RT60 Graphs',
            command=self.plot_combined
        )
        self.combine_button.grid(column=2, row=7, columnspan=2)

        self.combine_button.grid_forget()

        # Tkinter Waveform button
        self.waveform_graph_button = ttk.Button(
            self.root,
            text='Waveform Graph',
            command=self.plot_waveform
        )
        self.waveform_graph_button.grid(column=1, row=5, columnspan=2)

        self.waveform_graph_button.grid_forget()

        # Tkinter Intensity button
        self.intensity_graph_button = ttk.Button(
            self.root,
            text='Intensity Graph',
            command= self.plot_spectogram
        )
        self.intensity_graph_button.grid(column=0, row=5, columnspan=2)

        self.intensity_graph_button.grid_forget()
        
        # Canvas set up graph
        self.fig, self.ax = plt.subplots(figsize=(7, 4))
        self.ax.set_title('Default Graph')
        self.ax.set_xlabel('X-axis')
        self.ax.set_ylabel('Y-axis')
        self.canvas = FigureCanvasTkAgg(self.fig, master=self.root)
        self.canvas_widget = self.canvas.get_tk_widget()
        self.canvas_widget.grid(column=1, row=4, columnspan=2, padx=10, pady=10, sticky='nsew')


    # Swaps between RT60 graphs
    def swap_RT60_graphs(self):
        if self.count == 0:
            plt.clf()
            self.plot_low_rt60()
            self.count += 1
        elif self.count == 1:
            plt.clf()
            self.plot_middle_rt60()
            self.count += 1
        elif self.count == 2:
            plt.clf()
            self.plot_high_rt60()
            self.count = 0

    # Plots a waveform graph
    def plot_waveform(self):
        # Waveform Graph
        
        time_values = np.arange(0, len(self.model.data)) / self.model.samplerate
        
        plt.clf()
        plt.plot(time_values, self.model.data)
        plt.title('Waveform Graph')
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
      
        self.canvas.draw()
        
        
        length = round(self.model.data.shape[0] / self.model.samplerate, 2)
        self.length_label.config(text=f"File Length = {length}s")

    # Plots a spectogram graph
    def plot_spectogram(self):
        plt.clf()
        # Intensity Graph
        spectrum, freqs, t, im = plt.specgram(self.model.data, Fs=self.model.samplerate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
        cbar = plt.colorbar(im)
        cbar.set_label('Intensity (dB)')
        # drawing the intensity/spectogram graph
        plt.xlabel('Time (s)')
        plt.ylabel('Frequency (Hz)')
        plt.title('Frequency Graph')
        self.canvas.draw()

    # Calculates and plots the frequenciesRT60 Graph
    def plot_high_rt60(self):
        # RT60 High Graph
        self.plot_rt60("High", 1000, 5000, '#004bc6')
        
        self.canvas.draw()

    def plot_low_rt60(self):
        # RT60 Low Graph
        self.plot_rt60("Low", 20, 250, '#c67a00')
        
        self.canvas.draw()

    def plot_middle_rt60(self):
        # RT60 Mid Graph
        self.plot_rt60("Middle", 250, 1000, '#c600b6')
        
        self.canvas.draw()

    def plot_rt60(self, freq_range, low_freq, high_freq, color):
        # Find indices of frequencies within the specified range
        indices = np.where((self.model.freqs >= low_freq) & (self.model.freqs <= high_freq))[0]
        # Find the frequency with maximum intensity within the specified range
        self.target_frequency_index = indices[np.argmax(np.max(self.model.spectrum[indices, :], axis=1))]

        data_in_db = self.model.frequency_check(self.model.freqs[self.target_frequency_index])

        # plots a dot at the highest frequency
        plt.plot(self.model.t[self.target_frequency_index], data_in_db[self.target_frequency_index], 'go')

        # plotting a yellow dot at the -5db mark
        value_of_max_less_5 = data_in_db[self.target_frequency_index] - 5
        value_of_max_less_5 = self.model.find_nearest_value(data_in_db, value_of_max_less_5)
        index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)
        plt.plot(self.model.t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

        # plotting a red dot at the -25db mark
        value_of_max_less_25 = data_in_db[self.target_frequency_index] - 25
        value_of_max_less_25 = self.model.find_nearest_value(data_in_db, value_of_max_less_25)
        index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)
        plt.plot(self.model.t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')

        # calculating the RT20 value
        rt20 = (self.model.t[index_of_max_less_5] - self.model.t[index_of_max_less_25])[0]

        # calculating the RT60 value
        self.rt60 = 3 * rt20

        #Plot RT60
        plt.plot(self.model.t, data_in_db, alpha=0.7, color=color)
        plt.xlabel('Time (s)')
        plt.ylabel('Power (dB)')
        plt.title(f'{freq_range} RT60 Graph')


    def plot_combined(self):
        # RT60 Combined Graph
        plt.clf()
        self.plot_rt60("Combined", 1000, 5000, '#004bc6')
        self.plot_rt60("Combined", 20, 200, '#c67a00')
        self.plot_rt60("Combined", 200, 1000, '#c600b6')
        self.canvas.draw()

    def select_file(self):
        filetypes = (
            ('Sound files', '*.wav *.m4a *.aac *.mp3'),
            ('All files', '*.*')
        )

        filename = tk.filedialog.askopenfilename(
            title='Open a file',
            initialdir='/',
            filetypes=filetypes)

        self.gfile = filename
        
        justname = os.path.basename(filename)

        # Update the view
        self.gfile_label.config(text=f"File Name: {justname}")

        # Call the analyze_file method from the model
        self.analyze_file()

        # Tkinter Analyze button
        self.analyze_button = ttk.Button(
            self.root,
            text='Analyze File',
            command=self.analyze_file
        )
        self.analyze_button.grid(column=2, row=1)

    # Analyzes the file using methods from modle and controller
    def analyze_file(self):
        # Process the audio
        self.controller.process_audio(self.controller.output_path)
        # Call the analyze_file method from the model
        self.model.analyze_file(self.controller.output_path)

    # root.mainloop for Tkinter
    def mainloop(self):
            self.root.mainloop()
